src/cascadia/TerminalSettingsModel/FileUtils.cpp

```diff
@@ -0,0 +1,127 @@
+
+            // Create the directory if it doesn't exist
+            std::filesystem::create_directories(parentDirectoryForSettingsFile);
+
+            return parentDirectoryForSettingsFile;
+        }();
+        return baseSettingsPath;
+    }
+
+    std::string ReadUTF8File(const std::filesystem::path& path)
+    {
+        // From some casual observations we can determine that:
+        // * ReadFile() always returns the requested amount of data (unless the file is smaller)
+        // * It's unlikely that the file was changed between GetFileSize() and ReadFile()
+        // -> Lets add a retry-loop just in case, to not fail if the file size changed.
+        for (int i = 0; i < 3; ++i)
+        {
+            wil::unique_hfile file{ CreateFileW(path.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, nullptr) };
+            THROW_LAST_ERROR_IF(!file);
+
+            const auto fileSize = GetFileSize(file.get(), nullptr);
+            THROW_LAST_ERROR_IF(fileSize == INVALID_FILE_SIZE);
+
+            // By making our buffer just slightly larger we can detect if
+            // the file size changed and we've failed to read the full file.
+            std::string buffer(static_cast<size_t>(fileSize) + 1, '\0');
+            DWORD bytesRead = 0;
+            THROW_LAST_ERROR_IF(!ReadFile(file.get(), buffer.data(), gsl::narrow<DWORD>(buffer.size()), &bytesRead, nullptr));
+
+            // This implementation isn't atomic as it's not trivially possible to
+            // atomically read the entire file without breaking workflow of our users.
+            // For instance locking the file would conflict with users who have the file open in an editor.
+            // We'll just throw in case where the file size changed as that's likely an error.
+            if (bytesRead != fileSize)
+            {
+                // This continue is unlikely to be hit (see the prior for loop comment).
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                continue;
+            }
+
+            // As mentioned before our buffer was allocated oversized.
+            buffer.resize(bytesRead);
+
+            if (til::starts_with(buffer, Utf8Bom))
+            {
+                buffer.erase(0, Utf8Bom.size());
+            }
+
+            return buffer;
+        }
+
+        THROW_WIN32_MSG(ERROR_READ_FAULT, "file size changed while reading");
+    }
+
+    // Returns an empty optional (aka std::nullopt), if the file couldn't be opened.
+    // Use THROW_LAST_ERROR() if you want to fail in that case.
+    std::optional<std::string> ReadUTF8FileIfExists(const std::filesystem::path& path)
+    {
+        try
+        {
+            return { ReadUTF8File(path) };
+        }
+        catch (const wil::ResultException& exception)
+        {
+            if (exception.GetErrorCode() == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND))
+            {
+                return {};

```

Comment:
Yeah I know exceptions for control flow are bad, but I didn't have a good idea how to write this without being too roundabout.